### ISSUE TITLE
Using z.instanceof in z.union swallows invalid_union error

### ISSUE THREAD

**Issue by CGamesPlay** [colinhacks/zod#1140]
Given a schema like `z.union([z.string(), z.instanceof(Date)])`, you would expect `parse(null)` to return an `invalid_union` error. Instead of that, `custom` is returned from the `z.instanceof`. The following test case demonstrates:

```typescript
import { z } from "zod";

describe("zod", () => {
  it.each([
    z.union([z.string(), z.number()]),
    z.union([z.string(), z.instanceof(Date)]),
  ])("%#", (schema) => {
    const result = schema.safeParse(null);
    const error = (result as z.SafeParseError<any>).error;
    expect(error).toMatchObject({
      issues: [{ code: z.ZodIssueCode.invalid_union }],
    });
  });
});
```

The input can be any invalid value. The argument to `instanceof` can be any constructor. Using zod `3.16.0`.

**Comment by colinhacks** [comment]
What version of Zod are you using? On the latest version I can't replicate this.

```ts
z.union([z.string(), z.instanceof(Date)]).parse(null)
```
```
ZodError: [
  {
    "code": "invalid_union",
    "unionErrors": [
      {
        "issues": [
          {
            "code": "invalid_type",
            "expected": "string",
            "received": "null",
            "path": [],
            "message": "Expected string, received null"
          }
        ],
        "name": "ZodError"
      },
      {
        "issues": [
          {
            "code": "custom",
            "message": "Input not instance of Date",
            "fatal": true,
            "path": []
          }
        ],
        "name": "ZodError"
      }
    ],
    "path": [],
    "message": "Invalid input"
  }
]
```

**Comment by CGamesPlay** [comment]
Can confirm this was broken is `3.16.0` and fixed at/before `3.17.3`.
